guidelines/auth.md

```diff
@@ -0,0 +1,126 @@
+# Authentication and Authorization Guideline
+
+This guideline documents the current best practices for authentication and
+authorization in Kubeflow.
+* Kubeflow Application Developers should use this guide to effectively integrate
+  their applications with Kubeflow.
```

Comment:
Should we split this into two docs? These seem like very different audiences and topics. My suggestion would be to focus first on application developers as I believe one of the motivations for this doc was to ensure future app developers to comply.

It looks like there are two core questions this doc is trying to answer for developers.

1. How do application know the identity of the user making a request?
1. How do applications act on behalf of the user?

I would suggest focusing on those questions and trimming out some of the other stuff.

For example, as an application developer is how authentication performed really matter? Really all application developers need to know is that there will be a JWT and what fields in the JWT to look for.

Reply:
This doc is for app developers first and foremost, agreed.
I mentioned cluster admins as they may be able to get some useful info out of it.
The questions you enumerate are very accurate and I will include them explicitly in the doc.
I will try to trim based on that, however I think it's good to see a bit of the bigger picture before diving into the details.

> Really all application developers need to know is that there will be a JWT and what fields in the JWT to look for.

I'm a bit confused here. Currently all of our apps are looking for certain headers.
Is that what you mean?

Reply:
So if I recall correctly there are two bits of information that the app needs to know in order to be able to identify the user.

1. Which header field stores the JWT
1. How to parse the actual email from the email field
     * Because at least with IAP I think email ends up including a prefix like accounts.google.com

ref: https://cloud.google.com/iap/docs/signed-headers-howto

I may have gotten the details wrong. Goal is just to tell app developers how to parse the JWT

Reply:
Currently, our web applications don't parse any JWT.
They expect to find the user's identity in http headers.
This is what I describe in the doc.